### DISCUSSION TITLE
An element disappears upon ui.refreshable

### DISCUSSION THREAD

**Discussion by aapanchenko** [zauberzeug/nicegui#2898] [Q&A] [answered]
### Question

Code for Reproduction:

```py
@ui.refreshable
async def test_panael():
    # asynchronous HTTP request
    # state = async_httpx.get()
    # ui.label(state)
    await asyncio.sleep(2)
    ui.label(f'{datetime.now():%X}')

@ui.page('/')
async def test_refresh():
    await test_panael()
    ui.button('refresh time', on_click=test_panael.refresh)

ui.run()
```
Problem:
Date disappears when using asynchronous requests. However, everything works fine when using synchronous requests

**Comment by falkoschindler** [comment, marked as answer]
Hi @awfulness,

I would try to separate computation (HTTP requests in your case) and UI generation (the refreshable function). The `refresh()` function will clear the UI before running your code, so the UI is blank until `sleep(2)` has finished.

Here is an example how you can await the long-running functions first _before_ calling `refresh()`:

```py
@ui.refreshable
def test_panael(dt: datetime):
    ui.label(f'{dt:%X}')

async def handle_click():
    await asyncio.sleep(2)
    test_panael.refresh(datetime.now())

@ui.page('/')
async def test_refresh():
    test_panael(datetime.now())
    ui.button('refresh time', on_click=handle_click)
```

**Reply by aapanchenko** [reply]
In my non-toy example, this is not so easy to do. But I will try to follow this path, thank you!
